homeassistant/components/control4/__init__.py

```diff
@@ -0,0 +1,225 @@
+
+    director_token_dict = await account.getDirectorBearerToken(controller_unique_id)
+    director = C4Director(config[CONF_HOST], director_token_dict[CONF_TOKEN])
+    entry_data[CONF_DIRECTOR] = director
+    entry_data[CONF_DIRECTOR_TOKEN_EXPIRATION] = director_token_dict["token_expiration"]
+
+    # Add Control4 controller to device registry
+    controller_href = (await account.getAccountControllers())["href"]
+    entry_data[CONF_DIRECTOR_SW_VERSION] = await account.getControllerOSVersion(
+        controller_href
+    )
+
+    _, model, mac_address = controller_unique_id.split("_", 3)
+    entry_data[CONF_DIRECTOR_MODEL] = model.upper()
+
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        identifiers={(DOMAIN, controller_unique_id)},
+        connections={(dr.CONNECTION_NETWORK_MAC, mac_address)},
+        manufacturer="Control4",
+        name=controller_unique_id,
+        model=entry_data[CONF_DIRECTOR_MODEL],
+        sw_version=entry_data[CONF_DIRECTOR_SW_VERSION],
+    )
+
+    # Store all items found on controller for platforms to use
+    director_all_items = await director.getAllItemInfo()
+    director_all_items = json.loads(director_all_items)
+    entry_data[CONF_DIRECTOR_ALL_ITEMS] = director_all_items
+
+    # Load options from config entry
+    entry_data[CONF_SCAN_INTERVAL] = entry.options.get(
+        CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+    )
+
+    entry_data[CONF_CONFIG_LISTENER] = entry.add_update_listener(update_listener)
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def update_listener(hass, config_entry):
+    """Update when config_entry options update."""
+    _LOGGER.debug("Config entry was updated, rerunning setup")
+    await hass.config_entries.async_reload(config_entry.entry_id)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    hass.data[DOMAIN][entry.entry_id][CONF_CONFIG_LISTENER]()
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+        _LOGGER.debug("Unloaded entry for %s", entry.entry_id)
+
+    return unload_ok
+
+
+async def get_items_of_category(hass: HomeAssistant, entry: ConfigEntry, category: str):
+    """Return a list of all Control4 items with the specified category."""
+    director_all_items = hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR_ALL_ITEMS]
+    return_list = []
+    for item in director_all_items:
+        if "categories" in item.keys() and category in item["categories"]:
```

Comment:
We don't need to specify keys for membership test of dicts. That's the default. We can remove `.keys()`.